### PR TITLE
KAFKA-15307: Update/errors for deprecated config

### PR DESCRIPTION
**Pull request by Cerchie**
This PR is a follow-up on https://github.com/apache/kafka/pull/14360. 
It ensures that warnings are shown if the dep'd configs outlined in that PR are used (note topology.optimization's variable has changed rather than the configuration itself; this is already addressed in the code via KIP-626). 

Tests have been added to StreamsConfigTest.java to ensure the proper warnings show up. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by Cerchie** [comment]
tagging @mjsax for review

**Comment by mjsax** [comment]
Thanks for the PR. Overall LGTM.

I am just wondering if we should also update the JavaDocs for all deprecated configs? For example:
```
    /** {@code default.windowed.key.serde.inner} */
    @SuppressWarnings("WeakerAccess")
    @Deprecated
    public static final String DEFAULT_WINDOWED_KEY_SERDE_INNER_CLASS = "default.windowed.key.serde.inner";
```

We could change it to:
```
    /**
     * {@code default.windowed.key.serde.inner}
     * @deprecation Since X.y.0. Use {@link #WINDOWED_INNER_CLASS_SERDE} instead.
     */
    @SuppressWarnings("WeakerAccess")
    @Deprecated
    public static final String DEFAULT_WINDOWED_KEY_SERDE_INNER_CLASS = "default.windowed.key.serde.inner";
```

Thought? We could do this for _all_ deprecated configs?

**Comment by Cerchie** [comment]
> Thanks for the PR. Overall LGTM.
> 
> I am just wondering if we should also update the JavaDocs for all deprecated configs? For example:
> 
> ```
>     /** {@code default.windowed.key.serde.inner} */
>     @SuppressWarnings("WeakerAccess")
>     @Deprecated
>     public static final String DEFAULT_WINDOWED_KEY_SERDE_INNER_CLASS = "default.windowed.key.serde.inner";
> ```
> 
> We could change it to:
> 
> ```
>     /**
>      * {@code default.windowed.key.serde.inner}
>      * @deprecation Since X.y.0. Use {@link #WINDOWED_INNER_CLASS_SERDE} instead.
>      */
>     @SuppressWarnings("WeakerAccess")
>     @Deprecated
>     public static final String DEFAULT_WINDOWED_KEY_SERDE_INNER_CLASS = "default.windowed.key.serde.inner";
> ```
> 
> Thought? We could do this for _all_ deprecated configs?

went through and addressed all the missing ones I found

File: .github/workflows/codesee-arch-diagram.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# This workflow was added by CodeSee. Learn more at https://codesee.io/
```
Comment (mjsax) [inline .github/workflows/codesee-arch-diagram.yml]: I don't think this file belongs to this PR. Seems the tool you are using added it automatically...

File: docs/streams/developer-guide/config-streams.html [hunk 2]
```diff
@@ -91,6 +91,7 @@
               <li><a class="reference internal" href="#rocksdb-config-setter" id="id20">rocksdb.config.setter</a></li>
               <li><a class="reference internal" href="#state-dir" id="id14">state.dir</a></li>
               <li><a class="reference internal" href="#topology-optimization" id="id31">topology.optimization</a></li>
+              <li><a class="reference internal" href="#windowed-inner-class-serde" id="id38">windowed.inner.class.serde</a></li>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: As pointed out on the other PR. I think we should not add this config to the docs, but rather deprecate it similar to `window.size` via the KIP you started.

File: docs/streams/developer-guide/config-streams.html [hunk 3]
```diff
@@ -257,7 +258,12 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
             <td colspan="2">The maximum number of records to buffer per partition.</td>
             <td><code class="docutils literal"><span class="pre">1000</span></code></td>
           </tr>
-          <tr class="row-even"><td>cache.max.bytes.buffering</td>
+          <tr class="row-even"><td>statestore.cache.max.bytes</td>
+            <td>Medium</td>
+            <td colspan="2">Maximum number of memory bytes to be used for record caches across all threads. Note that at the debug level you can use <code>cache.size</code> to monitor the actual size of the Kafka Streams cache.</td>
+            <td>10485760</td>
+          </tr>
+          <tr class="row-even"><td>cache.max.bytes.buffering (Deprecated. Use cache.max.bytes instead.)</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:267]: even/odd issue...

File: docs/streams/developer-guide/config-streams.html [hunk 4]
```diff
@@ -300,14 +306,20 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
               set by the user or all serdes must be passed in explicitly (see also default.key.serde).</td>
             <td><code class="docutils literal"><span class="pre">null</span></code></td>
           </tr>
-          <tr class="row-even"><td>default.windowed.key.serde.inner</td>
+          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
             <td>Medium</td>
-            <td colspan="2">Default serializer/deserializer for the inner class of windowed keys, implementing the <code class="docutils literal"><span class="pre">Serde</span></code> interface.</td>
+<<<<<<< HEAD
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html]: Conflict not resolve correctly.

File: .github/workflows/codesee-arch-diagram.yml (hunk 1 above)
Comment (Cerchie) [reply .github/workflows/codesee-arch-diagram.yml]: removed

**Comment by Cerchie** [comment]
tagging @mjsax in for re-review

**Comment by mjsax** [comment]
@Cerchie -- this PR would need a rebase as #14360 was merged. Also, seems we introduced a few JavaDoc issues (cf https://github.com/apache/kafka/pull/14360#issuecomment-2108456039) that we should fix in this PR.

File: docs/streams/developer-guide/config-streams.html [hunk 5]
```diff
@@ -261,10 +261,10 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
           </tr>
           <tr class="row-odd"><td>statestore.cache.max.bytes</td>
             <td>Medium</td>
-            <td colspan="2">Maximum number of memory bytes to be used for record caches across all threads.</td>
+            <td colspan="2">Maximum number of memory bytes to be used for record caches across all threads. Note that at the debug level you can use <code>cache.size</code> to monitor the actual size of the Kafka Streams cache.</td>
             <td>10485760</td>
           </tr>
-          <tr class="row-even"><td>cache.max.bytes.buffering (Deprecated. Use statestore.cache.max.bytes instead.)</td>
+          <tr class="row-even"><td>cache.max.bytes.buffering (Deprecated. Use cache.max.bytes instead.)</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:267]: Seem `statestore.cache.max.bytes` is correct?

File: docs/streams/developer-guide/config-streams.html [hunk 6]
```diff
@@ -307,25 +307,23 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
...
             <td><code class="docutils literal"><span class="pre">null</span></code></td>
           </tr>
-          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
-            <td>Medium</td>
-            <td colspan="2">Default serializer/deserializer for the inner class of windowed keys, implementing the <code class="docutils literal"><span class="pre">Serde</span></code> interface.</td>
-            <td><code class="docutils literal"><span class="pre">null</span></code></td>
-          </tr>
-          <tr class="row-odd"><td>default.windowed.value.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
+          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated.)</td>
+
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:311]: nit: remove empty line

File: docs/streams/developer-guide/config-streams.html [hunk 7]
```diff
@@ -307,25 +307,23 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
               set by the user or all serdes must be passed in explicitly (see also default.key.serde).</td>
             <td><code class="docutils literal"><span class="pre">null</span></code></td>
           </tr>
-          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
-            <td>Medium</td>
-            <td colspan="2">Default serializer/deserializer for the inner class of windowed keys, implementing the <code class="docutils literal"><span class="pre">Serde</span></code> interface.</td>
-            <td><code class="docutils literal"><span class="pre">null</span></code></td>
-          </tr>
-          <tr class="row-odd"><td>default.windowed.value.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
+          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated.)</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:310]: Seems we don't need this line? It's duplicated above and below?

File: docs/streams/developer-guide/config-streams.html [hunk 8]
```diff
@@ -307,25 +307,23 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
...
+
+          <tr class="row-odd"><td>default.windowed.key.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:312]: duplicated line?

File: docs/streams/developer-guide/config-streams.html [hunk 9]
```diff
@@ -307,25 +307,23 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
               set by the user or all serdes must be passed in explicitly (see also default.key.serde).</td>
             <td><code class="docutils literal"><span class="pre">null</span></code></td>
           </tr>
-          <tr class="row-even"><td>default.windowed.key.serde.inner (Deprecated. Use windowed.inner.class.serde instead.)</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:310]: Why does this PR remove this row in the table?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 10]
```diff
@@ -500,7 +500,9 @@ public class StreamsConfig extends AbstractConfig {
     private static final String BUILT_IN_METRICS_VERSION_DOC = "Version of the built-in metrics to use.";
 
     /** {@code cache.max.bytes.buffering}
+
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:503]: nit: about unnecessary empty lines (same right below)

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 11]
```diff
@@ -658,7 +660,7 @@ public class StreamsConfig extends AbstractConfig {
     public static final String METRICS_SAMPLE_WINDOW_MS_CONFIG = CommonClientConfigs.METRICS_SAMPLE_WINDOW_MS_CONFIG;
 
     /** {@code auto.include.jmx.reporter}
-     * @deprecated and will be removed in 4.0.0 */
+     * @deprecated and will removed in 4.0.0 Users should instead include org.apache.kafka.common.metrics.JmxReporter in metric.reporters in order to enable the JmxReporter.} */
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:663]:
```suggestion
     * @deprecated and be will removed in 4.0.0; include {@link org.apache.kafka.common.metrics.JmxReporter} via config {@code metric.reporters} in order to enable the JmxReporter. */
```

File: docs/streams/developer-guide/config-streams.html [hunk 12]
```diff
@@ -261,10 +261,10 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
           </tr>
           <tr class="row-odd"><td>statestore.cache.max.bytes</td>
             <td>Medium</td>
-            <td colspan="2">Maximum number of memory bytes to be used for record caches across all threads.</td>
+            <td colspan="2">Maximum number of memory bytes to be used for record caches across all threads. Note that at the debug level you can use <code>cache.size</code> to monitor the actual size of the Kafka Streams cache.</td>
```
Comment (mjsax) [inline docs/streams/developer-guide/config-streams.html:264]: Not clear to me, what the added sentence means? Are you referring to JMX metrics?

**Comment by mjsax** [comment]
Closing is for now. We can reopen anytime.

**Pull request closed without merging**